src/chains/ethereum/ethereum/src/api.ts

```diff
@@ -2880,6 +2880,186 @@ export default class EthereumApi implements Api {
+      reward = new Array(totalBlocks);
+    }
+    // blockCount is inclusive of newestBlock
+    const oldestBlockNumber = newestBlockNumber - (totalBlocks - 1);
+    let currentBlockNumber = oldestBlockNumber;
+    let currentBlock: Block;
+
+    while (currentBlockNumber <= newestBlockNumber) {
+      currentBlock = await blockchain.blocks.get(
+        Quantity.toBuffer(currentBlockNumber)
+      );
+
+      const currentPosition = currentBlockNumber - oldestBlockNumber;
+
+      baseFeePerGas[currentPosition] =
+        currentBlock.header.baseFeePerGas || Quantity.Zero;
+
+      const { gasUsed, gasLimit } = currentBlock.header;
+      if (gasUsed === gasLimit) {
+        gasUsedRatio[currentPosition] = 1;
+      } else {
+        gasUsedRatio[currentPosition] = Number(
+          `0.${(
+            (currentBlock.header.gasUsed.toBigInt() * PRECISION_BIG_INT) /
+            currentBlock.header.gasLimit.toBigInt()
+          )
+            .toString()
+            .padStart(PAD_PRECISION, "0")}`
+        );
+      }
+
+      if (reward) {
+        const transactions = currentBlock.getTransactions();
+
+        // If there are no transactions, all reward percentiles are 0.
+        if (transactions.length === 0) {
+          reward[currentPosition] = rewardPercentiles.map(() => {
+            return Quantity.Zero;
+          });
+        } else {
+          // For all transactions, effectiveGasReward = normalized fee per unit of gas
+          // earned by the miner regardless of transaction type
+          const baseFeePerGas = currentBlock.header.baseFeePerGas
+            ? currentBlock.header.baseFeePerGas.toBigInt()
+            : 0n;
+
+          const effectiveRewardAndGasUsed = (
+            await Promise.all(
+              transactions.map(async tx => {
+                let effectiveGasReward: bigint;
+                if ("maxPriorityFeePerGas" in tx) {
+                  effectiveGasReward =
+                    tx.maxFeePerGas.toBigInt() - baseFeePerGas;
+
+                  const maxPriorityFeePerGas =
+                    tx.maxPriorityFeePerGas.toBigInt();
+
+                  if (maxPriorityFeePerGas < effectiveGasReward) {
+                    effectiveGasReward = maxPriorityFeePerGas;
+                  }
+                } else {
+                  effectiveGasReward = tx.gasPrice.toBigInt() - baseFeePerGas;
+                }
+
+                const receipt = (
+                  await blockchain.transactionReceipts.get(tx.hash.toBuffer())
+                ).toJSON(currentBlock, tx, blockchain.common);
+                return {
+                  effectiveGasReward: effectiveGasReward,
+                  gasUsed: receipt.gasUsed ? receipt.gasUsed.toBigInt() : 0n
+                };
+              })
+            )
+          ).sort((a, b) => {
+            if (a.effectiveGasReward > b.effectiveGasReward) return 1;
+            if (a.effectiveGasReward < b.effectiveGasReward) return -1;
+            return 0;
```

Comment:
This could become a thread.

Why insertion over a more stable sort like merge?  I agree that the more well-ordered the original list the closer to n runtime insertion becomes, but as the number of transactions increases with scale wouldn't the insertion sort algo drift towards n^2 where one like merge will always provide a stable nlog(n). 

I mean, what's an extra log(n) among friends?

So... memory constraints? Doing it in-place?

Reply:
Created: https://github.com/trufflesuite/ganache/issues/3804